File: src/BIOME4Py/phenology.py
"""Calculate the a generic phenology for any summergreen pft
three month period centred around the coldest month is
defined as the minimum period during which foliage is not
present. Plants then start growing leaves and the end of this
3 month period or when the temperature gos above 5oC if this
occurs later. Plants take 200 gdd5 to grow a full leaf canopy."""

# Third-party
import numpy as np


def phenology(
    dtemp: np.ndarray,
    temp: np.ndarray,
    tcm: float,
    tdif: float,
    tmin: float,
    pft: int,
    ddayl: np.ndarray,
    pftpar: np.ndarray,
) -> np.ndarray:
    daysinmonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    dphen = np.zeros((365, 2))
    ramp = [0.0, pftpar[pft][7], pftpar[pft][8]]

    if pft == 7:
        ont = 0.0  # minimum temp for growth
    else:
        ont = 5.0

    warm = tcm
    ncm = 0
    hotm = 0
    for m in range(12):
        if temp[m] == tcm:
            ncm = m + 1  # Fortran to Python index correction
        if temp[m] > warm:
            warm = temp[m]
            hotm = m + 1  # Fortran to Python index correction

    for phencase in range(2):
        coldm = [ncm - 1, ncm, ncm + 1]
        if coldm[0] == 0:
            coldm[0] = 12
        if coldm[2] == 13:
            coldm[2] = 1
        if hotm == 12:
            hotm = 0

        gdd = 0.0
        winter = 0
        for spinup in range(2):
            day = 0
            flip = 0
            for m in range(12):
                for dayofmonth in range(daysinmonth[m]):
                    day += 1
                    if dtemp[day - 1] > ont:
                        if m + 1 not in coldm:
                            today = max(dtemp[day - 1], 0.0)
                            gdd += today
                            if gdd == 0.0:
                                dphen[day - 1, phencase] = 0.0
                            else:
                                if ramp[phencase] != 0:
                                    dphen[day - 1, phencase] = gdd / ramp[phencase]
                                else:
                                    dphen[day - 1, phencase] = 0
                            if gdd >= ramp[phencase]:
                                dphen[day - 1, phencase] = 1.0
                            flip = 1
                        else:
                            if flip == 1:
                                winter = 0
                            winter += 1
                            dphen[day - 1, phencase] = 0.0
                            gdd = 0.0
                            flip = 0

                    if phencase == 1:
                        if m + 1 >= hotm:
                            if dtemp[day - 1] < -10.0 or ddayl[day - 1] < 10.0:
                                dphen[day - 1, phencase] = 0.0
                        elif m + 1 == coldm[0]:
                            dphen[day - 1, phencase] = 0.0
                    elif phencase == 2:
                        if dtemp[day - 1] < -5.0:
                            dphen[day - 1, phencase] = 0.0

    return dphen


if __name__ == "__main__":
    phenology()

File: src/BIOME4Ju/phenology.jl
"""Calculate the a generic phenology for any summergreen pft
three month period centred around the coldest month is
defined as the minimum period during which foliage is not
present. Plants then start growing leaves and the end of this
3 month period or when the temperature gos above 5oC if this
occurs later. Plants take 200 gdd5 to grow a full leaf canopy."""

module Phenology

using Base.Iterators: cycle

function phenology(
    dtemp::AbstractArray{Float64},
    temp::AbstractArray{Float64},
    tcm::Float64,
    tdif::Float64,
    tmin::Float64,
    pft::Int,
    ddayl::AbstractArray{Float64},
    pftpar::AbstractArray{Float64, 2}
)::AbstractArray{Float64}
    daysinmonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
    dphen = zeros(Float64, 365, 2)
    ramp = [0.0, pftpar[pft, 8], pftpar[pft, 9]]

    ont = pft == 7 ? 0.0 : 5.0

    warm = tcm
    ncm = 0
    hotm = 0
    for m in 1:12
        if temp[m] == tcm
            ncm = m
        end
        if temp[m] > warm
            warm = temp[m]
            hotm = m
        end
    end

    for phencase in 1:2
        coldm = [ncm - 1, ncm, ncm + 1]
        if coldm[1] == 0
            coldm[1] = 12
        end
        if coldm[3] == 13
            coldm[3] = 1
        end
        if hotm == 12
            hotm = 0
        end

        gdd = 0.0
        winter = 0
        for _ in 1:2
            day = 0
            flip = 0
            for m in 1:12
                for dayofmonth in 1:daysinmonth[m]
                    day += 1
                    if dtemp[day] > ont
                        if m ∉ coldm
                            today = max(dtemp[day], 0.0)
                            gdd += today
                            dphen[day, phencase] = if gdd == 0.0
                                0.0
                            elseif ramp[phencase] != 0
                                gdd / ramp[phencase]
                            else
                                0
                            end
                            if gdd >= ramp[phencase]
                                dphen[day, phencase] = 1.0
                            end
                            flip = 1
                        else
                            if flip == 1
                                winter = 0
                            end
                            winter += 1
                            dphen[day, phencase] = 0.0
                            gdd = 0.0
                            flip = 0
                        end
                    end

                    if phencase == 2
                        if m >= hotm
                            if dtemp[day] < -10.0 || ddayl[day] < 10.0
                                dphen[day, phencase] = 0.0
                            end
                        elseif m == coldm[1]
                            dphen[day, phencase] = 0.0
                        end
                    elseif phencase == 3
                        if dtemp[day] < -5.0
                            dphen[day, phencase] = 0.0
                        end
                    end
                end
            end
        end
    end

    return dphen
end

end # module

using .Phenology

# Example run
dtemp = [rand() * 20.0 for _ in 1:365]
temp = [10.0 for _ in 1:12]
tcm = -5.0
tdif = 0.0
tmin = -15.0
pft = 1
ddayl = [10.0 for _ in 1:365]
pftpar = rand(10, 10)

result = Phenology.phenology(dtemp, temp, tcm, tdif, tmin, pft, ddayl, pftpar)
println(result)
